### PR TITLE
Adding logic to support secrets from secrets manager(asm)

### PR DESCRIPTION
**Pull request by yumex93**
### Summary
Adding code change to support secrets from asm

### Implementation details
1. Add asm secret task resource
2. Add logic in task.go to initialize asm secrets
3. Add agent_capability for asm secrets vending as environment variable
4. Update the logic in createContainer in docker_task_engine.go to only populate secrets whose type are ENVIRONMENT_VARIABLE.

One thing wants to call out here, for asm, the valueFrom for secrets can be arn or name, while for ssm, it will only be name.

I carelessly delete the old branch so the old pr is closed. The link for old pr is here: https://github.com/aws/amazon-ecs-agent/pull/1669

### Testing
- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes
### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/docker_task_engine.go [hunk 1]
```diff
@@ -858,8 +858,8 @@ func (engine *DockerTaskEngine) createContainer(task *apitask.Task, container *a
 	}
 
 	// apply secrets to container.Environment
-	if container.ShouldCreateWithSSMSecret() {
```
Comment (haikuoliu) [inline agent/engine/docker_task_engine.go]: is `ShouldCreateWithSSMSecret` used somewhere else? if not we can remove it

File: agent/api/container/container.go [hunk 2]
```diff
@@ -775,8 +800,23 @@ func (c *Container) MergeEnvironmentVariables(envVars map[string]string) {
...
 		c.Environment = make(map[string]string)
 	}
-
 	for k, v := range envVars {
 		c.Environment[k] = v
 	}
 }
+
+func (c *Container) HasSecretAsEnv() bool {
+	c.lock.Lock()
```
Comment (sharanyad) [inline agent/api/container/container.go]: this can just be a read lock

File: agent/api/task/task.go [hunk 3]
```diff
@@ -1648,20 +1698,49 @@ func (task *Task) getASMAuthResource() ([]taskresource.TaskResource, bool) {
...
+		if secret.Provider == apicontainer.SecretProviderSSM && secret.Type == apicontainer.SecretTypeEnv {
+			k := secret.GetSecretResourceCacheKey()
+			if secretValue, ok := ssmRes.GetCachedSecretValue(k); ok {
+				envVars[secret.Name] = secretValue
+			}
+		}
+
+		if secret.Provider == apicontainer.SecretProviderASM && secret.Type == apicontainer.SecretTypeEnv {
+			k := secret.GetSecretResourceCacheKey()
+			if secretValue, ok := asmRes.GetCachedSecretValue(k); ok {
```
Comment (sharanyad) [inline agent/api/task/task.go]: Does this mean ASM takes precedence over SSM secret if both have same keys? Shall we also add a debug statement if we are overriding already added values?

File: agent/asm/asm.go [hunk 4]
```diff
@@ -86,3 +86,19 @@ func extractASMValue(out *secretsmanager.GetSecretValueOutput) (docker.AuthConfi
...
 }
+
+// GetSecretFromASM makes the api call to the AWS Secrets Manager service to
+// retrieve the secret value
+func GetSecretFromASM(secretID string, client secretsmanageriface.SecretsManagerAPI) (string, error) {
+	in := &secretsmanager.GetSecretValueInput{
+		SecretId: aws.String(secretID),
+	}
+
+	out, err := client.GetSecretValue(in)
```
Comment (sharanyad) [inline agent/asm/asm.go]: Does this have retries? Is there a possibility that customer would hit throttle errors here?

File: agent/taskresource/asmsecret/asmsecret.go [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+		return err
+	}
+	iamCredentials := executionCredentials.GetIAMRoleCredentials()
+
+	var wg sync.WaitGroup
+
+	// Get the maximum number of errors be returned, which will be one error per goroutine
+	errorEvents := make(chan error, len(secret.requiredSecrets))
+
+	seelog.Infof("asm secret resource: retrieving secrets for containers in task: [%s]", secret.taskARN)
```
Comment (sharanyad) [inline agent/taskresource/asmsecret/asmsecret.go]: nit: `ASM`

File: agent/taskresource/asmsecret/asmsecret.go [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+	wg.Wait()
+	close(errorEvents)
+
+	if len(errorEvents) > 0 {
+		var terminalReasons []string
+		for err := range errorEvents {
+			terminalReasons = append(terminalReasons, err.Error())
+		}
+
+		errorString := strings.Join(terminalReasons, ";")
```
Comment (sharanyad) [inline agent/taskresource/asmsecret/asmsecret.go]: 👍

File: agent/taskresource/asmsecret/asmsecret.go [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+	return nil
+}
+
+// retrieveASMSecretValue reads secret value from cache first, if not exists, call GetSecretFromASM to retrieve value
+// AWS secrets Manager
+func (secret *ASMSecretResource) retrieveASMSecretValue(apiSecret apicontainer.Secret, iamCredentials credentials.IAMRoleCredentials, wg *sync.WaitGroup, errorEvents chan error) {
+	defer wg.Done()
+
+	asmClient := secret.asmClientCreator.NewASMClient(apiSecret.Region, iamCredentials)
+	seelog.Infof("asm secret resource: retrieving resource for secret %v in region %s for task: [%s]", apiSecret.ValueFrom, apiSecret.Region, secret.taskARN)
```
[the remark this answers is not in the document]
Comment (yumex93) [reply agent/taskresource/asmsecret/asmsecret.go]: I want to differentiate it from ssm secret resource. I can update it to capital letters, however, I prefer to change both asm secret/ssm secret together with a separate commit later. Just to keep this pr clean.

File: agent/taskresource/asmsecret/asmsecret.go [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+		// No need to log here. managedTask.applyResourceState already does that
+		err := errors.New("asm secret resource: unable to find execution role credentials")
+		secret.setTerminalReason(err.Error())
+		return err
+	}
+	iamCredentials := executionCredentials.GetIAMRoleCredentials()
+
+	var wg sync.WaitGroup
+
+	// Get the maximum number of errors be returned, which will be one error per goroutine
```
[the remark this answers is not in the document]
Comment (yumex93) [reply agent/taskresource/asmsecret/asmsecret.go]: good catch

File: agent/api/task/task.go [hunk 9]
```diff
@@ -1732,6 +1732,58 @@ func (task *Task) getSSMSecretsResource() ([]taskresource.TaskResource, bool) {
...
+		asmRes = resource[0].(*asmsecret.ASMSecretResource)
+	}
+
+	envVars := make(map[string]string)
+
+	for _, secret := range container.Secrets {
+		if secret.Provider == apicontainer.SecretProviderSSM && secret.Type == apicontainer.SecretTypeEnv {
+			k := secret.GetSecretResourceCacheKey()
+			if secretValue, ok := ssmRes.GetCachedSecretValue(k); ok {
+				envVars[secret.Name] = secretValue
```
[the remark this answers is not in the document]
Comment (yumex93) [reply agent/api/task/task.go]: We do not validate whether multiple env variables have same names. We leave it to customers themselves to verify.

File: agent/api/container/container.go (hunk 2 above)
Comment (yumex93) [reply agent/api/container/container.go]: updated.

File: agent/api/task/task.go (hunk 3 above)
Comment (yumex93) [reply agent/api/task/task.go]: It is not really necessary to add a debug message here. This situation will also apply to secret values overriding normal environment variables. We do not check duplicated names on either ECS backend or agent side.  Leave it to customers to verify by themselves.

File: agent/asm/asm.go (hunk 4 above)
Comment (yumex93) [reply agent/asm/asm.go]: AWS SDK by default implements retry for their API calls. Customer may hit throttle error, however, they should be able to contact support to increase their limit.

File: agent/engine/docker_task_engine.go (hunk 1 above)
Comment (yumex93) [reply agent/engine/docker_task_engine.go]: There are multiple places using it. For example, line 537 in task.go

File: agent/api/container/container.go [hunk 10]
```diff
@@ -775,7 +797,7 @@ func (c *Container) MergeEnvironmentVariables(envVars map[string]string) {
 	if c.Environment == nil {
 		c.Environment = make(map[string]string)
 	}
-
+
```
[the remark this answers is not in the document]
Comment (yumex93) [reply agent/api/container/container.go]: I think you read the outdated version.

File: agent/taskresource/asmsecret/asmsecret.go (hunk 5 above)
Comment (yumex93) [reply agent/taskresource/asmsecret/asmsecret.go]: Updated

File: agent/taskresource/asmsecret/asmsecret.go [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+	errorEvents := make(chan error, len(secret.requiredSecrets))
```
[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/taskresource/asmsecret/asmsecret.go]: oh you are right, slice in golang is not thread safe

File: agent/api/task/task.go (hunk 3 above)
Comment (sharanyad) [reply agent/api/task/task.go]: > This situation will also apply to secret values overriding normal environment variables.  

Don't we do this check in backend and throw error message? 

I agree we can leave it to customers to make sure the keys are unique. But we are making a decision here to override the SSM secret values with ASM secret values if the keys are same. We should somehow let customers know this through docs and adding logs here might be useful.

**Pull request merged**